### PR TITLE
Create a no-op exposure runner

### PR DESCRIPTION
**Pull request by aranke**
### Problem

We want to make exposures executable with a no-op status in run results and log line, similar to saved queries.

### Solution

- [x] Create a new `NoOp` run status
- [x] Make `Exposure` extend `NodeInfoMixin` to make it executable in `build`
- [x] Create a new `NoOpRunner` in a new `dbt.runner` namespace, which uses the `NoOp` run status above
- [x] Make `SavedQueryRunner` inherit from `NoOpRunner` and move it to the `dbt.runner` namespace
- [x] Create a new `ExposureRunner` that also inherits from the `NoOpRunner`
- [x] Add new tests for run status for both exposures and saved queries
- [x] Add tests for selectors for exposures

#### Logs

```
❯ dbt build
12:39:45  Running with dbt=1.10.0-a1
12:39:45  Registered adapter: duckdb=1.9.1-dev28
12:39:46  Unable to do partial parsing because a project config has changed
12:39:46  Found 5 models, 3 seeds, 20 data tests, 1 exposure, 422 macros, 1 group
...
12:39:47  26 of 29 NO-OP exposure weekly_jaffle_metrics .................................. [NO-OP in 0.00s]
...
12:39:47  Finished running 1 exposure, 3 seeds, 2 table models, 20 data tests, 3 view models in 0 hours 0 minutes and 0.44 seconds (0.44s).
12:39:47  
12:39:47  Completed successfully
12:39:47  
12:39:47  Done. PASS=28 WARN=0 ERROR=0 SKIP=0 NO-OP=1 TOTAL=29
```

#### Run Results

```
❯ jq .results target/run_results.json
[
  ...
  {
    "status": "no-op",
    "timing": [
      {
        "name": "compile",
        "started_at": "2024-12-03T12:39:47.125472Z",
        "completed_at": "2024-12-03T12:39:47.125474Z"
      },
      {
        "name": "execute",
        "started_at": "2024-12-03T12:39:47.129151Z",
        "completed_at": "2024-12-03T12:39:47.129158Z"
      }
    ],
    "thread_id": "Thread-17 (worker)",
    "execution_time": 0.01341700553894043,
    "adapter_response": {},
    "message": "NO-OP",
    "failures": 0,
    "unique_id": "exposure.jaffle_shop.weekly_jaffle_metrics",
    "compiled": null,
    "compiled_code": null,
    "relation_name": null,
    "batch_results": null
  },
  ...
]
```

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me.
- [x] I have run this code in development, and it appears to resolve the stated issue.
- [x] This PR includes tests, or tests are not required or relevant for this PR.
- [x] This PR has no interface changes (e.g., macros, CLI, logs, JSON artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX.
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions.

### CODE REVIEW

File: core/dbt/contracts/graph/nodes.py [hunk 1]
```diff
@@ -1441,6 +1441,12 @@ def same_contents(self, old: Optional["Exposure"]) -> bool:
     def group(self):
         return None
 
+    def __post_serialize__(self, dct: Dict, context: Optional[Dict] = None):
+        dct = super().__post_serialize__(dct, context)
+        if "_event_status" in dct:
```
Comment (MichelleArk) [inline core/dbt/contracts/graph/nodes.py:1446]: why is this necessary?

File: core/dbt/runner/exposure_runner.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from dbt.runner.no_op_runner import NoOpRunner
```
Comment (MichelleArk) [inline core/dbt/runner/exposure_runner.py:1]: naming nit: let's rename this directory to `dbt/runners` (plural), similar to `dbt/tasks`

File: core/dbt/runner/saved_query_runner.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from dbt.runner.no_op_runner import NoOpRunner
+
+
+class SavedQueryRunner(NoOpRunner):
+    @property
+    def description(self) -> str:
+        return f"saved query {self.node.name}"
```
Comment (MichelleArk) [inline core/dbt/runner/saved_query_runner.py:7]: nice!

File: tests/functional/adapter/hooks/test_on_run_hooks.py [hunk 4]
```diff
@@ -125,7 +125,7 @@ def test_results(self, project):
         ]
 
         assert [(result.node.unique_id, result.status) for result in results] == expected_results
-        assert "PASS=3 WARN=0 ERROR=1 SKIP=1 TOTAL=5" in log_output
+        assert "PASS=3 WARN=0 ERROR=1 SKIP=1 NO-OP=0 TOTAL=5" in log_output
```
Comment (MichelleArk) [inline tests/functional/adapter/hooks/test_on_run_hooks.py:128]: because of test duplication, we'll need to update this in dbt-tests-adapters as well + cut a release to avoid breaking downstream concrete adapter tests: https://github.com/dbt-labs/dbt-adapters/blob/main/dbt-tests-adapter/dbt/tests/adapter/concurrency/test_concurrency.py#L329.

I'd opt for removing this assertion from the dbt-adapters tests. It seems brittle to test against core logs.

File: core/dbt/contracts/graph/nodes.py (hunk 1 above)
Comment (aranke) [reply core/dbt/contracts/graph/nodes.py]: You're right, this field is never written out.
Fixed the failing test to account for this field instead.

Comment (aranke) [reply core/dbt/contracts/graph/nodes.py:1446]: no sweat, just reverted the commit here

**Comment by MichelleArk** [comment]
Not strictly necessary to merge this PR, but let's just remember to make an update to the schemas in schemas.getdbt.com associated with this change.

**Review by MichelleArk — APPROVED** [review]

**Review by jtcohen6 — APPROVED** [review]
I left two comments, both of which confirm the behavior already present in this PR.

We manually confirmed that this should not run into the same issues we saw when adding hooks to the `on-run-end` `{{ results }}` context (#10699, #10885) because the structure of the no-op exposure result matches the structure of all other node results. Trying to access node-level properties that exist for other executable nodes, and do not exist for exposures (e.g. `schema`), will simply return a null value.

```
# dbt_project.yml
on-run-end:
  - "{% for result in results %} {{ log('The value of result.node.schema is: ' ~ result.node.schema, info = true) }} {% endfor %}"
```

🎩 Looks good:
```
% dbt retry
10:07:40  Running with dbt=1.10.0-a1
10:07:40  Registered adapter: duckdb=1.8.4
10:07:41  Found 1 seed, 1 operation, 1 model, 1 exposure, 421 macros
10:07:41  Nothing to do. Try checking your model configs and model specification args
(env) jerco@Jeremy-Cohen testy % dbt build -s my_model+
10:08:22  Running with dbt=1.10.0-a1
10:08:22  Registered adapter: duckdb=1.8.4
10:08:22  Unable to do partial parsing because a project config has changed
10:08:23  Found 1 model, 1 seed, 1 operation, 1 exposure, 421 macros
10:08:23  
10:08:23  Concurrency: 1 threads (target='dev')
10:08:23  
10:08:23  1 of 2 START sql table model main.my_model ..................................... [RUN]
10:08:23  1 of 2 ERROR creating sql table model main.my_model ............................ [ERROR in 0.04s]
10:08:23  2 of 2 SKIP exposure my_exposure ............................................... [SKIP]
10:08:23  
10:08:23  The value of result.node.schema for model.testy.my_model is: main
10:08:23  The value of result.node.schema for exposure.testy.my_exposure is: 
10:08:23  1 of 1 START hook: testy.on-run-end.0 .......................................... [RUN]
10:08:23  1 of 1 OK hook: testy.on-run-end.0 ............................................. [OK in 0.00s]
10:08:23  
10:08:23  Finished running 1 exposure, 1 project hook, 1 table model in 0 hours 0 minutes and 0.13 seconds (0.13s).
10:08:23  
10:08:23  Completed with 1 error, 0 partial successes, and 0 warnings:
10:08:23  
10:08:23    Runtime Error in model my_model (models/example/my_model.sql)
  Parser Error: syntax error at or near "bad_syntax"
10:08:23  
10:08:23  Done. PASS=1 WARN=0 ERROR=1 SKIP=1 NO-OP=0 TOTAL=3

# fix syntax error

% dbt retry             
10:08:52  Running with dbt=1.10.0-a1
10:08:52  Registered adapter: duckdb=1.8.4
10:08:52  Found 1 seed, 1 operation, 1 model, 1 exposure, 421 macros
10:08:52  
10:08:52  Concurrency: 1 threads (target='dev')
10:08:52  
10:08:52  1 of 2 START sql table model main.my_model ..................................... [RUN]
10:08:52  1 of 2 OK created sql table model main.my_model ................................ [OK in 0.04s]
10:08:52  2 of 2 NO-OP exposure my_exposure .............................................. [NO-OP in 0.00s]
10:08:52  
10:08:52  The value of result.node.schema for model.testy.my_model is: main
10:08:52  The value of result.node.schema for exposure.testy.my_exposure is: 
10:08:52  1 of 1 START hook: testy.on-run-end.0 .......................................... [RUN]
10:08:52  1 of 1 OK hook: testy.on-run-end.0 ............................................. [OK in 0.00s]
10:08:52  
10:08:52  Finished running 1 exposure, 1 project hook, 1 table model in 0 hours 0 minutes and 0.11 seconds (0.11s).
10:08:52  
10:08:52  Completed successfully
10:08:52  
10:08:52  Done. PASS=2 WARN=0 ERROR=0 SKIP=0 NO-OP=1 TOTAL=3

% dbt retry             
10:08:55  Running with dbt=1.10.0-a1
10:08:55  Registered adapter: duckdb=1.8.4
10:08:55  Found 1 seed, 1 operation, 1 model, 1 exposure, 421 macros
10:08:55  Nothing to do. Try checking your model configs and model specification args
```

In `run_results.json`:
```
{
            "status": "no-op",
            "timing": [
                {
                    "name": "compile",
                    "started_at": "2024-12-12T10:00:30.047130Z",
                    "completed_at": "2024-12-12T10:00:30.047131Z"
                },
                {
                    "name": "execute",
                    "started_at": "2024-12-12T10:00:30.047287Z",
                    "completed_at": "2024-12-12T10:00:30.047290Z"
                }
            ],
            "thread_id": "Thread-1",
            "execution_time": 0.0006129741668701172,
            "adapter_response": {},
            "message": "NO-OP",
            "failures": 0,
            "unique_id": "exposure.testy.my_exposure",
            "compiled": null,
            "compiled_code": null,
            "relation_name": null,
            "batch_results": null
        }
```

File: core/dbt/artifacts/schemas/results.py [hunk 5]
```diff
@@ -64,13 +64,15 @@ class NodeStatus(StrEnum):
...
     RuntimeErr = "runtime error"
+    NoOp = "no-op"
 
 
 class RunStatus(StrEnum):
     Success = NodeStatus.Success
     Error = NodeStatus.Error
     Skipped = NodeStatus.Skipped
     PartialSuccess = NodeStatus.PartialSuccess
+    NoOp = NodeStatus.NoOp
```
[the remark this answers is not in the document]
Comment (aranke) [reply core/dbt/artifacts/schemas/results.py:75]: I agree with this assessment because:
1. If an exposure has a failure/skipped status, it will be retried automatically (I will add a test for this case).
2. Its result won’t change unless the project config also changes (unlikely IME).

This is also a change we can implement fairly easily if we see a need for it, so I’ll merge the code in as-is for now.

File: core/dbt/graph/selector.py [hunk 6]
```diff
@@ -178,10 +178,12 @@ def _is_graph_member(self, unique_id: UniqueId) -> bool:
         elif unique_id in self.manifest.saved_queries:
             saved_query = self.manifest.saved_queries[unique_id]
             return saved_query.config.enabled
-
-        node = self.manifest.nodes[unique_id]
-
-        return node.config.enabled
+        elif unique_id in self.manifest.exposures:
```
Comment (QMalcolm) [inline core/dbt/graph/selector.py:181]: Adding a comment to create a link
